File: backend/vital_validation.py
"""Sanity validation for user-entered health measurements.

These are intentionally broad data-quality bounds, not clinical reference ranges.
They prevent malformed/impossible values from entering the medical history while
leaving clinical interpretation to the rest of the product.
"""

from __future__ import annotations

import math
from typing import Any, Dict, Optional, Tuple

from fastapi import HTTPException


RANGES: Dict[str, Tuple[float, float]] = {
    "systolic": (30.0, 350.0),
    "diastolic": (20.0, 250.0),
    "pulse": (20.0, 300.0),
    "weight": (0.1, 500.0),
    "temperature": (25.0, 45.0),
    "spo2": (1.0, 100.0),
    "waist": (10.0, 300.0),
}


def _finite_number(value: Any, field: str) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        raise HTTPException(422, f"Invalid {field}")
    if not math.isfinite(number):
        raise HTTPException(422, f"Invalid {field}")
    return number


def _bounded(value: Any, field: str) -> float:
    number = _finite_number(value, field)
    low, high = RANGES[field]
    if number < low or number > high:
        raise HTTPException(422, f"{field} is outside supported input range")
    return number


def validate_vital_payload(data: Any) -> None:
    """Validate a VitalCreate-like object before persistence."""

    kind = str(getattr(data, "kind", "") or "")
    if kind == "bp":
        systolic = _bounded(getattr(data, "systolic", None), "systolic")
        diastolic = _bounded(getattr(data, "diastolic", None), "diastolic")
        if systolic <= diastolic:
            raise HTTPException(422, "Systolic pressure must be greater than diastolic pressure")
        pulse: Optional[Any] = getattr(data, "pulse", None)
        if pulse is not None:
            _bounded(pulse, "pulse")
        return

    if kind not in {"weight", "temperature", "pulse", "spo2", "waist"}:
        raise HTTPException(422, "Unsupported vital kind")

    value = getattr(data, "value", None)
    _bounded(value, kind)

File: frontend/src/vitalValidation.ts
export type VitalKind = "weight" | "temperature" | "pulse" | "spo2" | "waist";

const ranges: Record<VitalKind | "systolic" | "diastolic", [number, number]> = {
  systolic: [30, 350],
  diastolic: [20, 250],
  pulse: [20, 300],
  weight: [0.1, 500],
  temperature: [25, 45],
  spo2: [1, 100],
  waist: [10, 300],
};

const parseNumber = (raw: string) => {
  const normalized = raw.trim().replace(",", ".");
  if (!normalized) return null;
  const value = Number(normalized);
  return Number.isFinite(value) ? value : null;
};

const inRange = (value: number, key: keyof typeof ranges) => {
  const [min, max] = ranges[key];
  return value >= min && value <= max;
};

export function validateBloodPressureInput(sysRaw: string, diaRaw: string, pulseRaw: string, ru: boolean) {
  const systolic = parseNumber(sysRaw);
  const diastolic = parseNumber(diaRaw);
  const pulse = pulseRaw.trim() ? parseNumber(pulseRaw) : null;
  const invalid = ru ? "Проверьте введённые значения" : "Check the entered values";

  if (systolic === null || diastolic === null) return { error: invalid } as const;
  if (!inRange(systolic, "systolic") || !inRange(diastolic, "diastolic")) return { error: invalid } as const;
  if (systolic <= diastolic) {
    return { error: ru ? "Верхнее давление должно быть выше нижнего" : "Systolic pressure must be higher than diastolic" } as const;
  }
  if (pulseRaw.trim() && (pulse === null || !inRange(pulse, "pulse"))) return { error: invalid } as const;

  return { value: { systolic, diastolic, pulse } } as const;
}

export function validateMeasurementInput(kind: VitalKind, raw: string, ru: boolean) {
  const value = parseNumber(raw);
  if (value === null || !inRange(value, kind)) {
    return { error: ru ? "Проверьте значение измерения" : "Check the measurement value" } as const;
  }
  return { value } as const;
}
